In [1]:
from nltk.corpus import wordnet

<b>1) Найти все значения (синсеты) для лексемы nail<b>

In [2]:
nail = wordnet.synsets('nail')
for ss in nail:
    print(ss, ss.definition())

Synset('nail.n.01') horny plate covering and protecting part of the dorsal surface of the digits
Synset('nail.n.02') a thin pointed piece of metal that is hammered into materials as a fastener
Synset('nail.n.03') a former unit of length for cloth equal to 1/16 of a yard
Synset('nail.v.01') attach something somewhere by means of nails
Synset('collar.v.01') take into custody
Synset('smash.v.01') hit hard
Synset('nail_down.v.01') succeed in obtaining a position
Synset('breeze_through.v.01') succeed at easily
Synset('pinpoint.v.01') locate exactly
Synset('complete.v.04') complete a pass


<b>2) Найти определение для лексемы nail в значении (а) "ноготь" и в значении (b) "гвоздь"</b>

In [3]:
print(nail[0], nail[0].definition())
print(nail[1], nail[1].definition())

Synset('nail.n.01') horny plate covering and protecting part of the dorsal surface of the digits
Synset('nail.n.02') a thin pointed piece of metal that is hammered into materials as a fastener


<b>3) Найдите два произвольных контекста для слова nail в значениях (а) "ноготь" и (b) "гвоздь"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [8]:
sent1 = "I looked down at my hands and picked at the dry skin around my nail."
sent2 = "One broad hand ripped his wet shirt over his head to hang on a nail."
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['I', 'looked', 'down', 'at', 'my', 'hands', 'and', 'picked', 'at', 'the', 'dry', 'skin', 'around', 'my', 'nail']
['One', 'broad', 'hand', 'ripped', 'his', 'wet', 'shirt', 'over', 'his', 'head', 'to', 'hang', 'on', 'a', 'nail']


In [9]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'nail').definition())
print(lesk(sent2_tokens, 'nail').definition())

horny plate covering and protecting part of the dorsal surface of the digits
a former unit of length for cloth equal to 1/16 of a yard


Алгоритм Леска ошибся во втором случае, приняв "гвоздь" за меру длины.

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [10]:
for ss in nail[0].hypernyms():
    print(ss, ss.definition())
for ss in nail[1].hypernyms():
    print(ss, ss.definition())

Synset('horny_structure.n.01') any rigid body structure composed primarily of keratin
Synset('fastener.n.02') restraint that attaches to something or holds something in place


<b>5) Вычислите наименьшее расстояние между значением nail "гвоздь" и значениями лексемы instrument, а также nail "ноготь" и значениями лексемы lunule.
Найти min (d(nail: "гвоздь", instrument), d(nail: "гвоздь", lunule)), а также min (d(nail: "ноготь", instrument), d(nail: "ноготь", lunule))</b>

In [13]:
instrument = wordnet.synsets('instrument')
lunule = wordnet.synsets('lunule')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(nail: "ноготь", instrument)
dist1 = get_dist_sim(nail[0], instrument)[0]
print('min d(nail: "ноготь", instrument): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(instrument[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(nail: "ноготь", lunule)
dist2 = get_dist_sim(nail[0], lunule)[0]
print('min d(nail: "ноготь", lunule): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(lunule[dist2.index(min(dist2))].definition()))

# min d(nail: "гвоздь", instrument)
dist3 = get_dist_sim(nail[1], instrument)[0]
print('min d(nail: "гвоздь", instrument): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(instrument[dist3.index(min(dist3))].definition()))

# min d(nail: "гвоздь", lunule)
dist4 = get_dist_sim(nail[1], lunule)[0]
print('min d(nail: "гвоздь", lunule): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(lunule[dist4.index(min(dist4))].definition()))

print('min (d(nail: "ноготь", instrument), d(nail: "ноготь", lunule)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(nail: "гвоздь", instrument), d(nail: "гвоздь", lunule)): {}'.format(min(min(dist3), min(dist4))))

min d(nail: "ноготь", instrument): 11
closest lemma definition: a person used by another to gain an end

min d(nail: "ноготь", lunule): 5
closest lemma definition: the crescent-shaped area at the base of the human fingernail

min d(nail: "гвоздь", instrument): 4
closest lemma definition: a device that requires skill for proper use

min d(nail: "гвоздь", lunule): 13
closest lemma definition: the crescent-shaped area at the base of the human fingernail

min (d(nail: "ноготь", instrument), d(nail: "ноготь", lunule)): 5
min (d(nail: "гвоздь", instrument), d(nail: "гвоздь", lunule)): 4


Кратчайшее расстояние от ногтя до инструмента - 11, причем ближайшее значение -- какое-то метафорическое, зато от ногтя до лунки - всего 5, и значение нужное. Кратчайшее расстояние от гвоздя до инструмента - 5, и значение верное. От гвоздя до лунки - 13, значение такое же как для ногтя (но видимо единственное у него).

Если имелось в виду посчитать не расстояние, а сходство путей на основе кратчайшего расстояния (минимальный путь --> максимальное сходство), вот оно:

In [14]:
# max s(nail: "ноготь", instrument)
sim1 = get_dist_sim(nail[0], instrument)[1]
print('max s(nail: "ноготь", instrument): {}'.format(max(sim1)))
print('closest lemma definition: {}\n'.format(instrument[sim1.index(max(sim1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# max s(nail: "ноготь", lunule)
sim2 = get_dist_sim(nail[0], lunule)[1]
print('max s(nail: "ноготь", lunule): {}'.format(max(sim2)))
print('closest lemma definition: {}\n'.format(lunule[sim2.index(max(sim2))].definition()))

# max s(nail: "гвоздь", instrument)
sim3 = get_dist_sim(nail[1], instrument)[1]
print('max s(nail: "гвоздь", instrument): {}'.format(min(sim3)))
print('closest lemma definition: {}\n'.format(instrument[sim3.index(max(sim3))].definition()))

# max s(nail: "гвоздь", lunule)
sim4 = get_dist_sim(nail[1], lunule)[1]
print('max s(nail: "гвоздь", lunule): {}'.format(max(sim4)))
print('closest lemma definition: {}\n'.format(lunule[sim4.index(max(sim4))].definition()))

print('max (s(nail: "ноготь", instrument), s(nail: "ноготь", lunule)): {}'.format(max(max(sim1), max(sim2))))
print('max (s(nail: "гвоздь", instrument), s(nail: "гвоздь", lunule)): {}'.format(max(max(sim3), max(sim4))))

max s(nail: "ноготь", instrument): 0.08333333333333333
closest lemma definition: a person used by another to gain an end

max s(nail: "ноготь", lunule): 0.16666666666666666
closest lemma definition: the crescent-shaped area at the base of the human fingernail

max s(nail: "гвоздь", instrument): 0.058823529411764705
closest lemma definition: a device that requires skill for proper use

max s(nail: "гвоздь", lunule): 0.07142857142857142
closest lemma definition: the crescent-shaped area at the base of the human fingernail

max (s(nail: "ноготь", instrument), s(nail: "ноготь", lunule)): 0.16666666666666666
max (s(nail: "гвоздь", instrument), s(nail: "гвоздь", lunule)): 0.2


Результат немного странный - у ногтя почему-то больше сходство с инструментом, хотя расстояние (пред. пункт) больше.

<b>6) Вычислить двумя разными способами расстояние:
d(nail: "ноготь", thumbnail) и d(organism, whole)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>

А вот здесь же точно речь идет о сходстве/близости значений, а не о стандартном расстоянии? (Потому что стандартное расстояние - кратчайший путь - трудно считать несколькими способами.)

In [15]:
thumbnail = wordnet.synsets("thumbnail")
for ss in thumbnail:
    print(ss, ss.definition())

Synset('thumbnail.n.01') the nail of the thumb


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [18]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(nail[0].path_similarity(thumbnail[0]))

#Leacock-Chodorow Similarity
print(nail[0].lch_similarity(thumbnail[0]))

#Wu-Palmer Similarity
print(nail[0].wup_similarity(thumbnail[0]))
    
#Resnik Similarity
print(nail[0].res_similarity(thumbnail[0], ic))
    
#Jiang-Conrath Similarity
print(nail[0].jcn_similarity(thumbnail[0], ic))
    
#Lin Similarity
print(nail[0].lin_similarity(thumbnail[0], ic))


0.3333333333333333
2.538973871058276
0.8888888888888888
11.330105697133238
0.318928988903801
0.8784489843654495


In [19]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [20]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [21]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
0.07692307692307693
1.072636802264849
0.14285714285714285
-0.0
0.11927453545173074
-0.0
an assemblage of parts that is regarded as a single entity
0.3333333333333333
2.538973871058276
0.8
1.5318337432196856
1.4444255924181877
0.8156763478002892
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
0.1
1.3350010667323402
0.3076923076923077
0.5962292078977726
0.05928363701403374
0.0660256991738309
an assemblage of parts that is regarded as a single entity
0.125
1.55814461804655
0.2222222222222222
-0.0
0.07444639262208605
-0.0


<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>ноготь / thumbnail</th><th>a living thing / all of something</th><th>a living thing / an assemblage</th><th>a system / all of something</th><th>a system / an assemblage</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.3333333333</td><td>0.07692307692</td><td>0.3333333333</td><td>0.1</td><td>0.125</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>2.538973871</td><td>1.072636802</td><td>2.538973871</td><td>1.335001067</td><td>1.558144618</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.8888888889</td><td>0.1428571429</td><td>0.8</td><td>0.3076923077</td><td>0.2222222222</td></tr>
 <tr><td>Resnik Similarity</td><td>11.3301056971</td><td>0</td><td>1.531833743</td><td>0.5962292079</td><td>0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>0.3189289889</td><td>0.1192745355</td><td>1.444425592</td><td>0.05928363701</td><td>0.07444639262</td></tr>
 <tr><td>Lin Similarity</td><td>0.8784489844</td><td>0</td><td>0.8156763478</td><td>0.06602569917</td><td>0</td></tr>
</tbody></table>

thumbnail является гипонимом ногтя. Как его сравнивать с organism/whole не совсем понятно, потому что во втором случае нет таких прямых отношений. За исключением предпоследнего индекса у этой пары самые большие показатели схожести.

Интуитивно кажется, что оба значения организма (a living thing that has (or can develop) the ability to act or function independently; a system considered analogous in structure or function to a living body) должны быть больше похожи на второе значение целого (an assemblage of parts that is regarded as a single entity), чем на первое (all of something including all its component elements or parts). То есть четвертый и шестой столбцы в таблице должны быть больше, чем третий и пятый. Хотя второе значение организма и первое значение целого тоже близки.

В четвертом столбце действительно самые большие индексы среди организма/целого. По предпоследнему индексу эта пара превосходит  ноготь / thumbnail, а по индексу Leacock-Chodorow Similarity равна ей. 

Показатели в шестом же столбце больше третьего и пятого только для первых двух индексов (по Wu-Palmer Similarity разница небольшая). Первые два индекса, кстати, дают более или менее пропорциональные величины, и сравнение по ним дает одинаковые результаты (это значит, что глубина наших значений в данном случае не влияет на результат).

Resnik Similarity и Lin Similarity в третьем и шестом столбце дают нули. Непонятно, говорит ли это о полной несхожести значений или это специфика индекса/корпуса/данных.

Если учитывать всё сразу, первые три индекса более или менее отражают интуитивное представление о близости слов.